hazelcast-tpc-engine/src/test/java/com/hazelcast/internal/tpcengine/net/RpcTest.java

```diff
@@ -0,0 +1,390 @@
+import com.hazelcast.internal.tpcengine.ReactorBuilder;
+import com.hazelcast.internal.tpcengine.iobuffer.IOBuffer;
+import com.hazelcast.internal.tpcengine.util.BufferUtil;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.hazelcast.internal.tpcengine.TpcTestSupport.ASSERT_TRUE_EVENTUALLY_TIMEOUT;
+import static com.hazelcast.internal.tpcengine.TpcTestSupport.assertJoinable;
+import static com.hazelcast.internal.tpcengine.TpcTestSupport.terminate;
+import static com.hazelcast.internal.tpcengine.net.AsyncSocketOptions.SO_RCVBUF;
+import static com.hazelcast.internal.tpcengine.net.AsyncSocketOptions.SO_SNDBUF;
+import static com.hazelcast.internal.tpcengine.net.AsyncSocketOptions.TCP_NODELAY;
+import static com.hazelcast.internal.tpcengine.util.BitUtil.SIZEOF_INT;
+import static com.hazelcast.internal.tpcengine.util.BitUtil.SIZEOF_LONG;
+
+/**
+ * Mimics an RPC call. So there are worker threads that send request with a call id and a payload. This request is
+ * 'processed' on the remote side and just returned.
+ * <p>
+ * On the local side a future is registered on that call id and when the response returns the future is looked up
+ * and notified. And then the worker thread will send another request.
+ */
+public abstract class RpcTest {
+    private static final int SIZEOF_HEADER = SIZEOF_INT + SIZEOF_LONG;
+
+    // use small buffers to cause a lot of network scheduling overhead (and shake down problems)
+    public static final int SOCKET_BUFFER_SIZE = 16 * 1024;
+    public long durationMillis = 500;
+    public long testTimeoutMs = ASSERT_TRUE_EVENTUALLY_TIMEOUT;
+    public boolean localWrite;
+    public boolean tcpNoDelay = true;
+    private final AtomicLong counter = new AtomicLong();
+    private final PrintAtomicLongThread printThread = new PrintAtomicLongThread("at:", counter);
+
+    private final ConcurrentMap<Long, CompletableFuture> futures = new ConcurrentHashMap<>();
+    private Reactor clientReactor;
+    private Reactor serverReactor;
+
+    public abstract ReactorBuilder newReactorBuilder();
+
+    protected void customizeClientSocketBuilder(AsyncSocketBuilder socketBuilder) {
+    }
+
+    protected void customizeServerSocketBuilder(AsyncSocketBuilder socketBuilder) {
+    }
+
+    @Before
+    public void before() {
+        clientReactor = newReactorBuilder().build().start();
+        serverReactor = newReactorBuilder().build().start();
+        printThread.start();
+    }
+
+    @After
+    public void after() throws InterruptedException {
+        terminate(clientReactor);
+        terminate(serverReactor);
+        printThread.shutdown();
+    }
+
+    @Test
```

Comment:
Can these tests be parameterized instead?

Reply:
Nope. Parametrization sucks if there is a problem and you just want to run explicit test.